File: jl_src/day_2.jl
import ArgParse

# Build ArgParsing Table
s = ArgParse.ArgParseSettings()
@ArgParse.add_arg_table s begin
    "--input"
        help = "Filepath to input data"
        arg_type = String
        required = true
    "--verbose"
        help = "Add additional reporting"
        action = :store_true
end

abstract type Location{T} end

mutable struct LocationA{T} <: Location{T}
    x::T
    depth::T
    LocationA{T}(x,y) where {T <: Number} = new(x,y)
end

mutable struct LocationB{T} <: Location{T}
    x::T
    depth::T
    aim::T
    LocationB{T}(x,y,z) where {T <: Number} = new(x,y,z)
end

abstract type Instruction{T} end

struct Down{T} <: Instruction{T}
    val::T
end

function execute(loc::LocationA{T}, inst::Down{T})::LocationA{T} where T <: Number
    return LocationA{T}(loc.x, loc.depth+inst.val)
end

function execute(loc::LocationB{T}, inst::Down{T})::LocationB{T} where T <: Number
    return LocationB{T}(loc.x, loc.depth, loc.aim+inst.val)
end

struct Up{T} <: Instruction{T}
    val::T
end

function execute(loc::LocationA{T}, inst::Up{T})::LocationA{T} where T <: Number
    return LocationA{T}(loc.x, loc.depth-inst.val)
end

function execute(loc::LocationB{T}, inst::Up{T})::LocationB{T} where T <: Number
    return LocationB{T}(loc.x, loc.depth, loc.aim-inst.val)
end

struct Forward{T} <: Instruction{T}
    val::T
end

function execute(loc::LocationA{T}, inst::Forward{T})::LocationA{T} where T <: Number
    return LocationA{T}(loc.x+inst.val, loc.depth)
end

function execute(loc::LocationB{T}, inst::Forward{T})::LocationB{T} where T <: Number
    return LocationB{T}(loc.x+inst.val, loc.depth+inst.val*loc.aim, loc.aim)
end

# Parse arguments
parsed_args = ArgParse.parse_args(ARGS, s)

struct ProcessFile{T} end

# Process files
function ProcessFile{T}(filepath::String)::Vector{Instruction{T}} where T<: Number
    result = Vector{Instruction{T}}()
    open(filepath) do f
        for line in readlines(f)
            m = match(r"(forward|up|down) ([0-9]*)", line)
            if m !== nothing
                inst = m.captures[1]
                val = parse(T, m.captures[2])
                if inst == "forward"
                    push!(result, Forward{T}(val))
                elseif inst == "up"
                    push!(result, Up{T}(val))
                elseif inst == "down"
                    push!(result, Down{T}(val))
                end
            end
        end
    end
    return result
end

# Get input filepath
input_filepath = parsed_args["input"]
verbose = parsed_args["verbose"]

# Process files
instructions = ProcessFile{Int32}(input_filepath)

function SolveDay1(instructions::Vector{Instruction{T}}) where T<: Number
    location = LocationA{T}(0,0)
    for inst in instructions
        location = execute(location, inst)
    end
    println("Day 2 problem 1: ", location.x*location.depth)
end

function SolveDay2(instructions::Vector{Instruction{T}}) where T<: Number
    location = LocationB{T}(0,0,0)
    for inst in instructions
        location = execute(location, inst)
    end
    println("Day 2 problem 2: ", location.x*location.depth)
end

SolveDay1(instructions)
SolveDay2(instructions)

File: rust_src/src/day_2.rs
extern crate argparse;

use std::io::{self, BufRead};
use std::fs::File;
use std::ops::{Add,Sub};
use std::str::FromStr;
use std::path::Path;
use argparse::{ArgumentParser, StoreTrue, Store};
use regex::Regex;

// Define readlines function discussed here:
// https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

struct Location<T>
where T: Add<Output = T>+Sub<Output = T>+Copy, {
    x: T,
    depth: T
}

enum Instructions<T>
where T: Add<Output = T>+Sub<Output = T>+Copy, {
    Forward(T),
    Down(T),
    Up(T)
}

impl<T> Location<T> where T: Add<Output = T>+Sub<Output = T>+Copy, {
    fn execute(&mut self, inst: &Instructions::<T>) {
        match inst {
            &Instructions::<T>::Forward(x) => {
                self.x = self.x + x;
            },
            &Instructions::<T>::Down(d) => {
                self.depth = self.depth + d;
            },
            &Instructions::<T>::Up(d) => {
                self.depth = self.depth - d;
            }
        }
    }
}

fn read_instructions<P, T>(filename: P) -> Vec<Instructions<T>>
where T: Add<Output = T>+Sub<Output = T>+Copy+FromStr, P: AsRef<Path>, {
    let re = Regex::new(r"(forward|down|up) ([0-9]*)").unwrap();
    let mut result = Vec::<Instructions::<T>>::new();
    // Reading the file
    if let Ok(lines) = read_lines(filename) {
        // Consumes the iterator, returns an (Optional) String
        for line in lines {
            if let Ok(ip) = line {
                for cap in re.captures_iter(&ip) {
                    let instruction = &cap[1];
                    if let Ok(val) = &cap[2].parse::<T>() {
                        match instruction {
                            "forward" => result.push(Instructions::<T>::Forward(*val)),
                            "up" => result.push(Instructions::<T>::Up(*val)),
                            "down" => result.push(Instructions::<T>::Down(*val)),
                            _ => ()
                        }
                    }
                }
            }
        }
    }
    result
}

fn main() {
    // Argument Parsing
    let mut verbose = false;
    let mut input = String::new();
    {
        let mut ap = ArgumentParser::new();
        ap.set_description("Day 2 AOC 2021");
        ap.refer(&mut input)
            .add_option(&["-i", "--input"], Store,
            "Filepath to input data");
        ap.refer(&mut verbose)
            .add_option(&["-v", "--verbose"], StoreTrue,
            "Add additional reporting");
        ap.parse_args_or_exit();
    }

    //read_instructions(input);
    let instructions: Vec::<Instructions::<i32>> = read_instructions(input);

    let mut location = Location::<i32>{
        x:0,
        depth:0
    };

    for inst in instructions {
        location.execute(&inst)
    }

    println!("Day 2 problem 1: {}", location.x*location.depth);
}
